src/validation.rs

```diff
@@ -78,12 +80,33 @@ impl<'s, T: AnnotationMap> ValidationContext<'s, T> {
             index: self.index,
             qualifier: self.qualifier,
             is_call: true,
+            is_builtin_call: self.is_builtin_call,
         }
     }
 
     fn is_call(&self) -> bool {
         self.is_call
     }
+
+    fn set_is_builtin_call(&self) -> Self {
+        ValidationContext {
+            annotations: self.annotations,
+            index: self.index,
+            qualifier: self.qualifier,
+            is_call: self.is_call,
+            is_builtin_call: true,
+        }
+    }
+
+    fn clone(&self) -> Self {
+        ValidationContext {
+            annotations: self.annotations,
+            index: self.index,
+            qualifier: self.qualifier,
+            is_call: self.is_call,
+            is_builtin_call: self.is_builtin_call,
+        }
+    }
```

Comment:
Deriving clone unfortunately doesn't work as expected due to the generic binding of `AnnotationMap`:
```
the method `clone` exists for struct `ValidationContext<'b, T>`, but its trait bounds were not satisfied
method cannot be called on `ValidationContext<'b, T>` due to unsatisfied trait bounds
```
Fully cloning the AnnotationMap is also not wanted in this case, so "cloning" refers to copying the references in this case. I am open to suggestions if someone can come up with a better name for this function.

Passing the validation context with a different context (e.g. `context.set_is_call()`) by reference into a recursive function call cause the compiler to complain about `returning a value referencing data owned by the current function` when collecting references from the index/annotator.